google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/BigQueryResultSetImpl.java

```diff
@@ -0,0 +1,554 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.bigquery;
+
+import com.google.cloud.Tuple;
+import java.math.BigDecimal;
+import java.sql.Date;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Time;
+import java.sql.Timestamp;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import java.time.ZoneId;
+import java.util.Map;
+import java.util.TimeZone;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.TimeUnit;
+import org.apache.arrow.vector.util.JsonStringArrayList;
+import org.apache.arrow.vector.util.Text;
+
+public class BigQueryResultSetImpl<T> implements BigQueryResultSet<T> {
+
+  private final Schema schema;
+  private final long totalRows;
+  private final BlockingQueue<T> buffer;
+  private T cursor;
+  private final ResultSetWrapper underlyingResultSet;
+  private final BigQueryResultSetStats bigQueryResultSetStats;
+  private final FieldList schemaFieldList;
+
+  public BigQueryResultSetImpl(
+      Schema schema,
+      long totalRows,
+      BlockingQueue<T> buffer,
+      BigQueryResultSetStats bigQueryResultSetStats) {
+    this.schema = schema;
+    this.totalRows = totalRows;
+    this.buffer = buffer;
+    this.underlyingResultSet = new ResultSetWrapper();
+    this.bigQueryResultSetStats = bigQueryResultSetStats;
+    this.schemaFieldList = schema.getFields();
+  }
+
+  @Override
+  public Schema getSchema() {
+    return schema;
+  }
+
+  @Override
+  public long getTotalRows() {
+    return totalRows;
+  }
+
+  @Override
+  public ResultSet getResultSet() {
+    return underlyingResultSet;
+  }
+
+  private class ResultSetWrapper extends AbstractJdbcResultSet {
+    @Override
+    /*Advances the result set to the next row, returning false if no such row exists. Potentially blocking operation*/
+    public boolean next() throws SQLException {
+      try {
+        cursor = buffer.take(); // advance the cursor,Potentially blocking operation
+        if (isEndOfStream(cursor)) { // check for end of stream
+          cursor = null;
+          return false;
+        } else if (cursor instanceof Tuple) {
+          Tuple<Map<String, Object>, Boolean> curTup = (Tuple<Map<String, Object>, Boolean>) cursor;
+          if (!curTup.y()) { // last Tuple
+            cursor = null;
+            return false;
+          }
+        }
+      } catch (InterruptedException e) {
+        throw new SQLException("Error occurred while reading buffer");
```

Comment:
makes sense, update the message to `Error occurred while advancing the cursor. This could happen when connection is closed while we call the next method`